crates/uv-python/src/installation.rs

```diff
@@ -92,36 +92,48 @@ impl PythonInstallation {
         let request = request.unwrap_or(&PythonRequest::Default);
 
         // Search for the installation
-        match Self::find(request, environments, preference, cache) {
-            Ok(venv) => Ok(venv),
-            // If missing and allowed, perform a fetch
-            Err(Error::MissingPython(err))
-                if preference.allows_managed()
-                    && python_downloads.is_automatic()
-                    && client_builder.connectivity.is_online() =>
-            {
-                if let Some(request) = PythonDownloadRequest::from_request(request) {
-                    debug!("Requested Python not found, checking for available download...");
-                    match Self::fetch(
-                        request.fill()?,
-                        client_builder,
-                        cache,
-                        reporter,
-                        python_install_mirror,
-                        pypy_install_mirror,
-                    )
-                    .await
-                    {
-                        Ok(installation) => Ok(installation),
-                        Err(Error::Download(downloads::Error::NoDownloadFound(_))) => {
-                            Err(Error::MissingPython(err))
-                        }
-                        Err(err) => Err(err),
-                    }
-                } else {
-                    Err(Error::MissingPython(err))
-                }
-            }
+        let err = match Self::find(request, environments, preference, cache) {
+            Ok(installation) => return Ok(installation),
+            Err(err) => err,
+        };
+
+        let downloads_enabled = preference.allows_managed()
+            && python_downloads.is_automatic()
+            && client_builder.connectivity.is_online();
+
+        if !downloads_enabled {
+            return Err(err);
+        }
+
+        match err {
+            // If Python is missing, we should attempt a download
+            Error::MissingPython(_) => {}
+            // If we raised a non-critical error, we should attempt a download
+            Error::Discovery(ref err) if !err.is_critical() => {}
```

Comment:
This is the key change. The rest is just refactoring so we can match on multiple error types cleanly.